### PR TITLE
[WFLY-9403] Unable to deploy clustered ejb in a mixed domain

### PR DESCRIPTION
**Pull request by yersan**
A tentative solution for WFLY-9403, I still need some feedback on this issue, see my comments.

Issue: https://issues.jboss.org/browse/WFLY-9403
JBEAP Issue: https://issues.jboss.org/browse/JBEAP-13295

### CODE REVIEW

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java [hunk 1]
```diff
@@ -301,4 +320,59 @@ private static void registerPassivationStoreTransformers_1_2_1_and_1_3_0(Resourc
...
+                operation = operation.clone();
+
+                String defaultSfsbCache = operation.get(DEFAULT_SFSB_CACHE).asString();
+                String defaultSfsbPassivationDisabledCache = operation.get(DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE).asString();
+
+                operation.get(DEFAULT_SFSB_CACHE).set(defaultSfsbPassivationDisabledCache);
+                operation.get(EJB3SubsystemModel.DEFAULT_CLUSTERED_SFSB_CACHE).set(defaultSfsbCache);
+                operation.remove(DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE);
+
+            } else if ( !(operation.hasDefined(OPERATION_HEADERS) && operation.get(OPERATION_HEADERS, "push-to-servers").asBoolean(false)) ){
```
Comment (yersan) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: I cannot evaluate if the use of this header is the correct solution. There is a corner case with the transformation required for this issue. What we have to do is implement the following transformation:
cache-ref --> clustered-cache-ref
passivation-disabled-cache-ref --> cache-ref

Here, the custom operation transformer is invoked more than once for different addresses to transform the same operation. There are two code paths which trigger this request of transformation:

1. Operations pushed into the HC
https://github.com/wildfly/wildfly-core/blob/46ad87495026705fd559698d32069978b38f7071/host-controller/src/main/java/org/jboss/as/domain/controller/operations/coordination/HostControllerUpdateTask.java#L97

2. Operations pushed into the Server groups pushed by the HC
https://github.com/wildfly/wildfly-core/blob/46ad87495026705fd559698d32069978b38f7071/host-controller/src/main/java/org/jboss/as/domain/controller/operations/coordination/DomainRolloutStepHandler.java#L355

The operation transformation triggered by 2) is invoked for an operation already transformed by the invocation 1). That situation is the corner case for this issue because if we are transforming `passivation-disabled-cache-ref` into `cache-ref`, we will end up transforming again from `cache-ref` to `clustered-cache-ref` due to 2).

The presence of the header "push-to-servers" in the invocation 2) seems to solve this dilemma, but I'm not sure if this is the correct solution.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java [hunk 2]
```diff
@@ -301,4 +320,59 @@ private static void registerPassivationStoreTransformers_1_2_1_and_1_3_0(Resourc
...
+
+        public SfsbChacheRefTransformer(){
+            renames = new HashMap<>();
+            renames.put(DEFAULT_SFSB_CACHE, EJB3SubsystemModel.DEFAULT_CLUSTERED_SFSB_CACHE);
+            renames.put(DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE, EJB3SubsystemModel.DEFAULT_SFSB_CACHE);
+        }
+
+        @Override
+        public TransformedOperation transformOperation(TransformationContext context, PathAddress address, ModelNode operation) throws OperationFailedException {
+            if (Operations.getName(operation).equals(ADD)){
```
Comment (yersan) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: The only requirement I have found to add here the ADD operation is due to the use of the transform in the test. I could find out a case wherein a mixed domain we have to add the EJB3 subsystem, maybe if as a user I remove the whole subsystem to add it later. Do you know another example that requires this operation added to the transform?

Comment (yersan) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: This operation needs also take care of "push-to-servers" header in order to work fine. I will update the PR

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java [hunk 3]
```diff
@@ -301,4 +321,70 @@ private static void registerPassivationStoreTransformers_1_2_1_and_1_3_0(Resourc
...
+
+        public SfsbChacheRefTransformer(){
+            renames = new HashMap<>();
+            renames.put(DEFAULT_SFSB_CACHE, EJB3SubsystemModel.DEFAULT_CLUSTERED_SFSB_CACHE);
+            renames.put(DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE, EJB3SubsystemModel.DEFAULT_SFSB_CACHE);
+        }
+
+        @Override
+        public TransformedOperation transformOperation(TransformationContext context, PathAddress address, ModelNode operation) throws OperationFailedException {
+            if (operation != null && !(operation.hasDefined(OPERATION_HEADERS) && operation.get(OPERATION_HEADERS, "push-to-servers").asBoolean(false)) ){
```
Comment (yersan) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: I cannot evaluate if the use of this header is the correct solution. There is a corner case with the transformation required for this issue. What we have to do is implement the following transformation:
cache-ref --> clustered-cache-ref
passivation-disabled-cache-ref --> cache-ref

Here, the custom operation transformer is invoked more than once for different addresses to transform the same operation. There are two code paths which trigger this request of transformation:

1. Operations pushed into the HC
    https://github.com/wildfly/wildfly-core/blob/46ad87495026705fd559698d32069978b38f7071/host-controller/src/main/java/org/jboss/as/domain/controller/operations/coordination/HostControllerUpdateTask.java#L97

2. Operations pushed into the Server groups pushed by the HC
    https://github.com/wildfly/wildfly-core/blob/46ad87495026705fd559698d32069978b38f7071/host-controller/src/main/java/org/jboss/as/domain/controller/operations/coordination/DomainRolloutStepHandler.java#L355

The operation transformation triggered by 2) is invoked for an operation already transformed by the invocation 1). That situation is the corner case for this issue because if we are transforming `passivation-disabled-cache-ref` into `cache-ref`, we will end up transforming again from `cache-ref` to `clustered-cache-ref` due to 2).

The presence of the header "push-to-servers" in the invocation 2) seems to solve this dilemma, but I'm not sure if this is the correct solution.

Comment (bstansberry) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: I think this is an ok workaround. I filed https://issues.jboss.org/browse/WFCORE-3439 to look into why this double transformation happens.

**Comment by bstansberry** [comment]
@pferraro The transformation looks ok to me. Is the testing of this adequate?

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java [hunk 4]
```diff
@@ -301,4 +321,70 @@ private static void registerPassivationStoreTransformers_1_2_1_and_1_3_0(Resourc
...
     }
+
+    /**
+     * This Combined Transformer manages this transformation from EAP7 to previous versions (attribute definition/xml markup):
+     *
+     * DEFAULT_SFSB_CACHE / cache-ref --> DEFAULT_CLUSTERED_SFSB_CACHE / clustered-cache-ref
+     * DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE / passivation-disabled-cache-ref --> DEFAULT_SFSB_CACHE / cache-ref
+     *
+     */
+    private static class SfsbChacheRefTransformer implements CombinedTransformer {
```
Comment (pferraro) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: Can we fix the spelling?  "StatefulCacheRefTransformer" is probably clearer.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java [hunk 5]
```diff
@@ -301,4 +321,70 @@ private static void registerPassivationStoreTransformers_1_2_1_and_1_3_0(Resourc
...
+            transformed.get(DEFAULT_SFSB_CACHE).set(defaultSfsbPassivationDisabledCache);
+            transformed.get(EJB3SubsystemModel.DEFAULT_CLUSTERED_SFSB_CACHE).set(defaultSfsbCache);
+            transformed.remove(DEFAULT_SFSB_PASSIVATION_DISABLED_CACHE);
+
+            final ResourceTransformationContext childContext = context.addTransformedResource(PathAddress.EMPTY_ADDRESS, resource);
+            childContext.processChildren(resource);
+        }
+    }
+
+    private static class AddSfsbChacheRefTransformer implements OperationTransformer{
```
Comment (pferraro) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: Same "chache" spelling problem.  Kinda reminds me of ehcache. :)

File: ejb3/src/test/resources/org/jboss/as/ejb3/subsystem/subsystem-ejb3-transform.xml [hunk 6]
```diff
@@ -3,7 +3,7 @@
         <stateless>
             <bean-instance-pool-ref pool-name="slsb-strict-max-pool"/>
         </stateless>
-        <stateful default-access-timeout="${prop.default-access-timeout:5000}" cache-ref="distributable" clustered-cache-ref="distributable"/>
+        <stateful default-access-timeout="${prop.default-access-timeout:5000}" cache-ref="distributable" clustered-cache-ref="distributable" passivation-disabled-cache-ref="simple"/>
```
Comment (pferraro) [inline ejb3/src/test/resources/org/jboss/as/ejb3/subsystem/subsystem-ejb3-transform.xml]: We need logic added to Ejb3TransformersTestCase that validates this transformation.  Can you add this?

**Review by pferraro — CHANGES_REQUESTED** [review]

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java (hunk 4 above)
Comment (yersan) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: @pferraro I've fixed the spelling problem.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java (hunk 5 above)
Comment (yersan) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java]: @pferraro I've fixed the spelling problem.

File: ejb3/src/test/resources/org/jboss/as/ejb3/subsystem/subsystem-ejb3-transform.xml (hunk 6 above)
Comment (yersan) [reply ejb3/src/test/resources/org/jboss/as/ejb3/subsystem/subsystem-ejb3-transform.xml]: @pferraro Sorry, would you mind being more specific about what do you need exactly? As far as I understand, transformer test framework is already testing this attribute. 
The transformer test framework creates the model for the target version using the operation transformer. Later the model is validated using the resource transformer, if the models match, then the transformation is supposed to be fine. So, in essence, if we include the three attributes in the subsystem xml, we will test the rename for all of them. WDYT?

I would like to add a small note here that might require a closer look, it looks like we are forced to include `passivation-disabled-cache-ref="simple"` in this scenario, I mean, the scenario where we are using a subsystem configuration including the deprecated attribute `clustered-cache-ref="distributable"`.

By default, if we use `clustered-cache-ref`, its value is used to initialize the `passivation-disabled-cache-ref` https://github.com/wildfly/wildfly/blob/341bce7d37617fbad19a5e2b44dc60fff9b383fd/ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemAdd.java#L208

I think this is also a corner case for the test framework, because if we don't define `passivation-disabled-cache-ref` in this scenario, the test will never pass due to how test framework is validating the models; matching the model created with the operation transform against the model created with the resource transform.

A different solution would be to use only this pair of attributes: `cache-ref="distributable" passivation-disabled-cache-ref="simple"`, but doing so, we won't be testing the rename using all attributes, that's the reason I included all of them.

Comment (pferraro) [reply ejb3/src/test/resources/org/jboss/as/ejb3/subsystem/subsystem-ejb3-transform.xml]: @yersan What I mean is, I don't think there is anything logic in Ejb3TransformersTestCase that specifically checks that the transformation results for these attributes is correct - only that the transformation results in a valid/consistent model.

Comment (yersan) [reply ejb3/src/test/resources/org/jboss/as/ejb3/subsystem/subsystem-ejb3-transform.xml]: Hi @pferraro, I have updated the test coverage to test the transformation executing write/read/undefine operations, let me know what do you think.

**Review by pferraro — APPROVED** [review]

**Comment by pferraro** [comment]
Looks good.  Nice work!
